### PR TITLE
Upgrade to cypari2

### PR DESCRIPTION
**Pull request by saraedum**

### CODE REVIEW

**Comment by saraedum** [comment]
I would like to package this for conda-forge. However, cypari is not on conda-forge but cypari2 is which is probably better anyway…

**Comment by saraedum** [comment]
cypari vendors pari it seems. So some changes are needed to make the CI happy again.

If you are in general fine with this, I can try to fix the CI.

**Comment by videlec** [comment]
cypari2 is not an upgrade of cypari but a fork. The main reason why cypari still exists is because
- it supports Windows (now, cypari2 should also thanks to [cysignals #63](https://github.com/sagemath/cysignals/issues/63))
- it provides binary wheels which package all together: gmp, cysignals, pari and the Python interface (see [cypari2 #19](https://github.com/sagemath/cypari2/issues/19)). This simplifies the installation procedure for users (that do not have a distribution).

**Comment by saraedum** [comment]
Thanks for the clarification. So, what about making realalg work with cypari & cypari2, whichever is available?

**Comment by videlec** [comment]
I think this would be reasonable (especially if people are using SageMath which is bundled with cypari2). However it will be delicate to make the `setup.py` script of realalg to handle the dependencies correctly. How do you specify alternative dependencies?

**Comment by saraedum** [comment]
I don't think this can be done. The sane approach seems to not have an explicit dependency in `setup.py` but throw an exception if neither is installed that tells you what you should do.

**Comment by videlec** [comment]
Much less user friendly than an explicit dependency...

**Comment by saraedum** [comment]
I don't disagree but unfortunately setup.py has no hooks to do something like this. The other option is to split this into two pypi packages with different dependencies but I don't think that this is very helpful either.

**Comment by videlec** [comment]
@MarkCBell  Because of #3 it would be nice to allow both cypari or cypari2.

(BTW, I expect https://github.com/flatsurf/pyeantic (mostly written in C) to be much faster...)

**Comment by videlec** [comment]
In azure and appveyor Python 3.5, 3.6 and 3.7 we got
```
py35 run-test: commands[0] | py.test --hypothesis-profile=ci --duration=20 --junitxml=results.xml
ERROR: usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: --duration=20
  inifile: /home/vsts/work/1/s/tox.ini
  rootdir: /home/vsts/work/1/s

ERROR: InvocationError for command /home/vsts/work/1/s/.tox/py35/bin/py.test --hypothesis-profile=ci --duration=20 --junitxml=results.xml (exited with code 4)
```

File: realalg/algebraic.py [hunk 1]
```diff
@@ -6,13 +6,21 @@
...
 import sympy as sp
+try:
+    import cypari2
+    cp = cypari2.Pari()
+except ImportError:
+    try:
+        import cypari
+        cp = cypari.pari
+    except ImportError:
+        raise Exception("You need to install cypari or cypari2 to use the realalg module. Try pip install cypari.")
```
Comment (videlec) [inline realalg/algebraic.py]: I would rather advice `pip install cypari2` rather than `cypari`.

**Comment by videlec** [comment]
The test failure has nothing to do with @saraedum branch, see #4

**Comment by saraedum** [comment]
@MarkCBell @videlec tests pass now.

**Comment by videlec** [comment]
See also the related https://github.com/MarkCBell/flipper/issues/3

File: realalg/algebraic.py [hunk 2]
```diff
@@ -6,13 +6,21 @@
 from math import log10 as log
 from numbers import Integral
 
-import cypari as cp
 import sympy as sp
+try:
```
Comment (videlec) [inline realalg/algebraic.py]: Independently of the installation procedure, this `try/except` in the code is an improvement!

**Comment by MarkCBell** [comment]
Sorry for the delay but I think I finally have a solution for this. I've now written a number of interfaces to backends that use cypari, cypari2 and sympy. This means that the only required dependency is now sympy. So if used on its own then realalg will fall back on its sympy implementation which is good enough to allow users to run basic examples (slowly). However if it can import cypari or cypari2 then it will use these, allowing larger more complex examples to be done.

The latest version of realalg (0.2.2) has this functionality. Let me know what you think.

**Comment by videlec** [comment]
Very good! Thanks.

You can definitely close this merge request.

**Pull request closed without merging**